File: qr-check-latestcode/QR_CHECK_IN/code/app/src/main/java/com/example/qr_check_in/Notification/Permission.java
package com.example.qr_check_in.Notification;
import android.os.Build;

import androidx.annotation.RequiresApi;

public abstract class Permission {
    private String[] permissions;

    private Permission(String... permissions) {
        this.permissions = permissions;
    }

    public static final Permission Camera = new Permission("CAMERA") {};
    public static final Permission MandatoryForFeatureOne = new Permission("WRITE_EXTERNAL_STORAGE", "ACCESS_FINE_LOCATION") {};
    public static final Permission Location = new Permission("ACCESS_FINE_LOCATION", "ACCESS_COARSE_LOCATION") {};
    public static final Permission Storage = new Permission("WRITE_EXTERNAL_STORAGE", "READ_EXTERNAL_STORAGE") {};

    @RequiresApi(33)
    public static final Permission Notification = new Permission("POST_NOTIFICATIONS") {};

    public static Permission from(String permission) {
        switch (permission) {
            case "ACCESS_FINE_LOCATION":
            case "ACCESS_COARSE_LOCATION":
                return Location;
            case "WRITE_EXTERNAL_STORAGE":
            case "READ_EXTERNAL_STORAGE":
                return Storage;
            case "POST_NOTIFICATIONS":
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    return Notification;
                }
            case "CAMERA":
                return Camera;
            default:
                throw new IllegalArgumentException("Unknown permission: " + permission);
        }
    }
}

File: DmcRemAlert/app/src/main/java/com/example/dmcremalert/Permission.kt
package com.example.dmcremalert

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.util.Log
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.app.ActivityCompat
import androidx.fragment.app.FragmentActivity
import com.example.dmcremalert.navigators.AddPermissions
import com.example.dmcremalert.viewmodel.DeviceListViewModel


class Permission(activity: FragmentActivity, var context: Context, private var viewModel: DeviceListViewModel):
    AddPermissions {
    private val BLE_PERMISSIONS = arrayOf<String>(
        android.Manifest.permission.ACCESS_COARSE_LOCATION,
        android.Manifest.permission.ACCESS_FINE_LOCATION
    )

    private val ANDROID_12_BLE_PERMISSIONS = arrayOf<String>(
        android.Manifest.permission.BLUETOOTH_SCAN,
        android .Manifest.permission.BLUETOOTH_CONNECT,
        android.Manifest.permission.ACCESS_FINE_LOCATION,
        Manifest.permission.WRITE_EXTERNAL_STORAGE,
        Manifest.permission.BLUETOOTH_ADMIN,
        Manifest.permission.WRITE_EXTERNAL_STORAGE


        )

    override fun requestBlePermissions(activity: Activity?, requestCode: Int) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) multiplePermissionLauncher!!.launch(
            ANDROID_12_BLE_PERMISSIONS
        ) else multiplePermissionLauncher!!.launch(
            BLE_PERMISSIONS
        )
    }

    private val multiplePermissionsContract: ActivityResultContracts.RequestMultiplePermissions =
        ActivityResultContracts.RequestMultiplePermissions()

    private val multiplePermissionLauncher: ActivityResultLauncher<Array<String>>? =
        activity?.registerForActivityResult(multiplePermissionsContract) { grantResults ->
            val isGranted: Boolean = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                java.lang.Boolean.TRUE == grantResults.get(Manifest.permission.ACCESS_FINE_LOCATION) || java.lang.Boolean.TRUE == grantResults.get(
                    Manifest.permission.BLUETOOTH_SCAN
                ) || java.lang.Boolean.TRUE == grantResults.get(Manifest.permission.BLUETOOTH_ADMIN) || java.lang.Boolean.TRUE == grantResults.get(
                    Manifest.permission.BLUETOOTH_CONNECT
                )
            } else {
                java.lang.Boolean.TRUE == grantResults.get(Manifest.permission.ACCESS_FINE_LOCATION) || java.lang.Boolean.TRUE == grantResults.get(
                    Manifest.permission.BLUETOOTH_ADMIN
                )
            }
            if (isGranted) {
                Log.d("granted","granted")
                viewModel.startScanning()
            } else {
                Manifest.permission()
            }
        }

     override fun permission() {
         if (multiplePermissionLauncher != null) {
             askPermissions(multiplePermissionLauncher)
         }
    }

    override fun askPermissions(multiplePermissionLauncher: ActivityResultLauncher<Array<String>>) {
        if (!hasPermissions(ANDROID_12_BLE_PERMISSIONS)) {
            Log.d(
                "PERMISSIONS", "Launching multiple contract permission launcher " +
                        "for ALL required permissions"
            )
            multiplePermissionLauncher.launch(ANDROID_12_BLE_PERMISSIONS)
        } else {
           viewModel.startScanning()
            Log.d("PERMISSIONS", "All permissions are already granted")
        }
    }

   override  fun hasPermissions(permissions: Array<String>?): Boolean {
        if (permissions != null) {
            for (permission in permissions) {
                if (ActivityCompat.checkSelfPermission(context, permission)
                    != PackageManager.PERMISSION_GRANTED
                ) {
                    Log.d("PERMISSIONS", "Permission is not granted: $permission")
                    return false
                }
               Log.d("PERMISSIONS", "Permission already granted: $permission")
            }
            return true
        }
        return false
    }



}